src/ir_parser.c

```diff
@@ -0,0 +1,1297 @@
+            } break;
+
+            /// <binary> <temp> "," <temp>
+            /// <binary> ::= ADD | SUB | MUL | DIV | MOD | EQ | NE | LT | LE | GT | GE | SHL | SHR | SAR
+            case ADD:
+            case SUBTRACT:
+            case MULTIPLY:
+            case DIVIDE:
+            case MODULO:
+            case EQ:
+            case NE:
+            case LT:
+            case LE:
+            case GT:
+            case GE:
+            case SHIFT_LEFT:
+            case SHIFT_RIGHT_LOGICAL:
+            case SHIFT_RIGHT_ARITHMETIC: {
+                next_token(p);
+
+                /// Parse the first temporary.
+                if (p->tok_type != tk_temp) ERR_AT(i_loc, "Expected temporary after binary instruction");
+                IRInstruction *a = resolve_curr_temp(p);
+                next_token(p);
+
+                /// Yeet ",".
+                if (p->tok_type != tk_comma) ERR_AT(i_loc, "Expected ',' after first temporary in binary instruction");
+                next_token(p);
+
+                /// Parse the second temporary.
+                if (p->tok_type != tk_temp) ERR_AT(i_loc, "Expected temporary after ',' in binary instruction");
+                IRInstruction *b = resolve_curr_temp(p);
+                next_token(p);
+
+                /// Create the instruction.
+                switch (type) {
+                    case ADD: i = ir_add(p->context, a, b); break;
+                    case SUBTRACT: i = ir_subtract(p->context, a, b); break;
+                    case MULTIPLY: i = ir_multiply(p->context, a, b); break;
+                    case DIVIDE: i = ir_divide(p->context, a, b); break;
+                    case MODULO: i = ir_modulo(p->context, a, b); break;
+                    case EQ: i = ir_comparison(p->context, COMPARE_EQ, a, b); break;
+                    case NE: i = ir_comparison(p->context, COMPARE_NE, a, b); break;
+                    case LT: i = ir_comparison(p->context, COMPARE_LT, a, b); break;
+                    case LE: i = ir_comparison(p->context, COMPARE_LE, a, b); break;
+                    case GT: i = ir_comparison(p->context, COMPARE_GT, a, b); break;
+                    case GE: i = ir_comparison(p->context, COMPARE_GE, a, b); break;
+                    case SHIFT_LEFT: i = ir_shift_left(p->context, a, b); break;
+                    case SHIFT_RIGHT_LOGICAL: i = ir_shift_right_logical(p->context, a, b); break;
+                    case SHIFT_RIGHT_ARITHMETIC: i = ir_shift_right_arithmetic(p->context, a, b); break;
+                    default: UNREACHABLE();
+                }
+            } break;
+
+            /// LOAD ( <temp> | <name> )
+            case LOAD: {
+                next_token(p);
+
+                /// Parse the temporary or name.
+                if (p->tok_type == tk_temp) i = ir_load(p->context, resolve_curr_temp(p));
+                else if (p->tok_type == tk_ident) i = ir_load_global(p->context, strndup(p->tok.data, p->tok.size));
```

Comment:
1. `strndup` is a POSIX function and only integrated into C in C23. It doesn't exist on Windows.
`#define strndup(data, size) strdup(data)`
2. `strndup` is defined in `<string.h>` which isn't included.

Reply:
`strdup()` would be a *very* bad idea here since we’re explicitly *not* dealing w/ null-terminated strings here. I’m just gonna implement `strndup()` myself.